applications/FluidDynamicsApplication/custom_processes/shock_capturing_entropy_viscosity_process.cpp

```diff
@@ -0,0 +1,390 @@
+    const Element& rElement,
+    const double DeltaTime,
+    const double HeatCapacityRatio)
+{
+    TotalDerivativeUtil entropy_td;
+    TotalDerivativeUtil density_td;
+    Vector total_velocities;
+
+    std::tie(entropy_td, density_td, total_velocities) = BuildTotalDerivativeUtils(rElement, DeltaTime, HeatCapacityRatio);
+    return ComputeInfNorms(rElement.GetGeometry(), entropy_td, density_td, total_velocities);
+}
+
+
+std::tuple<ShockCapturingEntropyViscosityProcess::TotalDerivativeUtil, ShockCapturingEntropyViscosityProcess::TotalDerivativeUtil, Vector>
+ShockCapturingEntropyViscosityProcess::BuildTotalDerivativeUtils(const Element& rElement, const double DeltaTime, const double HeatCapacityRatio)
+{
+    KRATOS_TRY
+
+    const auto& r_geometry = rElement.GetGeometry();
+
+    // Loading nodal values
+    TotalDerivativeUtil entropy_total_derivative(r_geometry.LocalSpaceDimension(), r_geometry.size());
+    TotalDerivativeUtil density_total_derivative(r_geometry.LocalSpaceDimension(), r_geometry.size());
+    Vector total_velocities(r_geometry.size(), 0.0);
+
+    for(unsigned int i=0; i<r_geometry.size(); ++i)
+    {
+        const auto& r_node = r_geometry[i];
+
+        const auto velocity = r_node.FastGetSolutionStepValue(VELOCITY);
+        const auto temperature = r_node.FastGetSolutionStepValue(TEMPERATURE);
+
+        total_velocities[i] = norm_2(velocity) + std::sqrt(HeatCapacityRatio * temperature);
+
+        entropy_total_derivative.LoadNodalValues(NUMERICAL_ENTROPY, r_node, i, velocity, DeltaTime);
+        density_total_derivative.LoadNodalValues(DENSITY, r_node, i, velocity, DeltaTime);
+        entropy_total_derivative.Value[i] /= density_total_derivative.Value[i];
+    }
+
+    // Computing inf norm over gauss points
+    return std::tie(entropy_total_derivative, density_total_derivative, total_velocities);
+
+    KRATOS_CATCH("")
+}
+
+
+ShockCapturingEntropyViscosityProcess::InfNormData ShockCapturingEntropyViscosityProcess::ComputeInfNorms(
+    const Geometry<NodeType>& rGeometry,
+    const TotalDerivativeUtil& rEntropyTotalDerivative,
+    const TotalDerivativeUtil& rDensityTotalDerivative,
+    const Vector& rTotalVelocities)
+{
+    KRATOS_TRY
+
+    constexpr auto integration_method = GeometryData::IntegrationMethod::GI_GAUSS_1;
+    const auto n_gauss_points = rGeometry.IntegrationPointsNumber(integration_method);
+
+    const auto& r_shape_functions = rGeometry.ShapeFunctionsValues(integration_method);
+    GeometryData::ShapeFunctionsGradientsType r_shape_functions_gradients;
+    rGeometry.ShapeFunctionsIntegrationPointsGradients(r_shape_functions_gradients, integration_method);
+
+    double max_residual = 0.0;  // max(Dh1, Dh2) in the paper
+    double max_density = 0.0;
+    double max_total_velocity = 0.0;
+
+    for(unsigned int g = 0; g<n_gauss_points; ++g)
+    {
+        const auto N = row(r_shape_functions, g);
```

Comment:
You could create an auxiliary array outside the loop to avoid allocating each time you retrieve the values.

Reply:
There are no dynamic allocations in this loop.

The only suspect is boost's `row` but this is a reference to the matrix it is extracted from, so no allocations are made (I wasn't sure so I tested it: https://godbolt.org/z/hrd39anE8).

Reply:
Cool. I thought `row` was creating a copy. Thanks for the testing.

Reply:
(I'm not resolving as this might be interesting for other people as well)